server/upsd.c

```diff
@@ -275,8 +277,169 @@ static void setuptcp(stype_t *server)
 	struct addrinfo		hints, *res, *ai;
 	int	v = 0, one = 1;
 
+	if (VALID_FD_SOCK(server->sock_fd)) {
+		/* Alredy bound, e.g. thanks to 'LISTEN *' handling and injection
+		 * into the list we loop over */
+		upsdebugx(6, "setuptcp: SKIP bind to %s port %s: entry already initialized",
+			server->addr, server->port);
+		return;
+	}
+
 	upsdebugx(3, "setuptcp: try to bind to %s port %s", server->addr, server->port);
 
+	/* Special handling for `LISTEN * <port>` directive with literal asterisk:
+	 * on systems with RFC-3493 (no relation!) support for "IPv4-mapped addresses"
+	 * it suffices to LISTEN on "::" (aka "::0" or "0:0:0:0:0:0:0:0") and also
+	 * get an IPv4 any-address listener. More so, they would conflict and
+	 * listening on one such socket precludes listening on the other. On other
+	 * systems (or with disabled mapping so IPv6 means IPv6 only) we need both.
+	 * So we jump through some hoops:
+	 * * Try to get IPv4 any-address, just to know if it is available right now;
+	 * * Free it and try to get IPv6 any-address, and try to get again that
+	 *   IPv4 any-address (IFF it was available before but is not available now -
+	 *   not a problem).
+	 * * Remember the entries used, to release later.
+	 * For more details see https://github.com/networkupstools/nut/issues/2012
+	 */
```

Comment:
meta: I don't think it's good to leave github issue refs in the code.  In the glorious future, nut will migrate away and everyone else too and github will fade away.   But seriously, comments in the code base should explain enough, and the issue should not be necessary (and thus not appear).

Reply:
Well, fair point but over the years quite a few such references got there. On the upside, GitHub REST API makes it possible to keep backups... FWIW, I began this post hours ago, but it led me on a detour to update https://github.com/jimklimov/github-scripts with better handling for issues/pulls backups :-D